### PR TITLE
register sysview to note drivers list

### PR DESCRIPTION
**Pull request by Gary-Hobson**
## Summary
register sysview to note drivers list

## Impact
still only one note driver can be used
multi-channel functionality needs to wait for sysview changes to complete

## Testing
sim:note

### CODE REVIEW

File: drivers/segger/sysview.c [hunk 1]
```diff
@@ -406,7 +420,7 @@ void sched_note_syscall_enter(int nr, int argc, ...)
   SEGGER_SYSVIEW_MarkStart(nr);
 }
 
-void sched_note_syscall_leave(int nr, uintptr_t result)
+static void sysview_syscall_leave(FAR struct note_driver_s *drv, int nr)
 {
   nr -= CONFIG_SYS_RESERVED;
```
Comment (xiaoxiang781216) [inline drivers/segger/sysview.c:426]: remove sysview_isenabled_syscall and NOTE_FILTER_SYSCALLMASK_ISSET, already check in common code

File: drivers/segger/sysview.c [hunk 2]
```diff
@@ -371,7 +385,7 @@ void sched_note_irqhandler(int irq, FAR void *handler, bool enter)
 #endif
 
 #ifdef CONFIG_SCHED_INSTRUMENTATION_SYSCALL
-void sched_note_syscall_enter(int nr, int argc, ...)
+static void sysview_syscall_enter(FAR struct note_driver_s *drv, int nr)
 {
   nr -= CONFIG_SYS_RESERVED;
```
Comment (xiaoxiang781216) [inline drivers/segger/sysview.c:391]: remove sysview_isenabled_syscall and NOTE_FILTER_SYSCALLMASK_ISSET, already check in common code

File: drivers/segger/sysview.c [hunk 3]
```diff
@@ -333,7 +346,8 @@ void sched_note_resume(FAR struct tcb_s *tcb)
 #endif
 
 #ifdef CONFIG_SCHED_INSTRUMENTATION_IRQHANDLER
-void sched_note_irqhandler(int irq, FAR void *handler, bool enter)
+static void sysview_irqhandler(FAR struct note_driver_s *drv, int irq,
+                               FAR void *handler, bool enter)
 {
   if (!sysview_isenabled_irq(irq, enter))
```
Comment (xiaoxiang781216) [inline drivers/segger/sysview.c:352]: remove, already check in common code

File: drivers/segger/sysview.c [hunk 4]
```diff
@@ -311,7 +323,8 @@ void sched_note_suspend(FAR struct tcb_s *tcb)
     }
 }
 
-void sched_note_resume(FAR struct tcb_s *tcb)
+static void sysview_resume(FAR struct note_driver_s *drv,
+                           FAR struct tcb_s *tcb)
 {
   if (!sysview_isenabled())
```
Comment (xiaoxiang781216) [inline drivers/segger/sysview.c:329]: remove

File: drivers/segger/sysview.c [hunk 5]
```diff
@@ -298,7 +309,8 @@ void sched_note_stop(FAR struct tcb_s *tcb)
 }
 
 #ifdef CONFIG_SCHED_INSTRUMENTATION_SWITCH
-void sched_note_suspend(FAR struct tcb_s *tcb)
+static void sysview_suspend(FAR struct note_driver_s *drv,
+                            FAR struct tcb_s *tcb)
 {
   if (!sysview_isenabled())
```
Comment (xiaoxiang781216) [inline drivers/segger/sysview.c:315]: remove

File: drivers/segger/sysview.c [hunk 6]
```diff
@@ -287,7 +297,8 @@ void sched_note_start(FAR struct tcb_s *tcb)
   sysview_send_taskinfo(tcb);
 }
 
-void sched_note_stop(FAR struct tcb_s *tcb)
+static void sysview_stop(FAR struct note_driver_s *drv,
+                         FAR struct tcb_s *tcb)
 {
   if (!sysview_isenabled())
```
Comment (xiaoxiang781216) [inline drivers/segger/sysview.c:303]: remove

File: drivers/segger/sysview.c [hunk 7]
```diff
@@ -276,7 +285,8 @@ static inline int sysview_isenabled_syscall(int nr)
  *
  ****************************************************************************/
 
-void sched_note_start(FAR struct tcb_s *tcb)
+static void sysview_start(FAR struct note_driver_s *drv,
+                          FAR struct tcb_s *tcb)
 {
   if (!sysview_isenabled())
```
Comment (xiaoxiang781216) [inline drivers/segger/sysview.c:291]: remove

File: drivers/segger/sysview.c [hunk 8]
```diff
@@ -71,6 +124,15 @@ static struct sysview_s g_sysview =
 #endif
 };
 
+/****************************************************************************
+ * Public Data
+ ****************************************************************************/
+
+struct note_driver_s g_sysview_driver =
```
Comment (xiaoxiang781216) [inline drivers/segger/sysview.c:131]: embed into sysview_s

File: drivers/segger/sysview.c [hunk 9]
```diff
@@ -54,10 +53,64 @@ struct sysview_s
 #endif
```
Comment (xiaoxiang781216) [inline drivers/segger/sysview.c:53]: remove mode, irq_mask and syscall_mask

Comment (xiaoxiang781216) [reply drivers/segger/sysview.c:53]: sysview_s->note_sysview_driver_s

File: drivers/segger/sysview.c (hunk 2 above)
Comment (Gary-Hobson) [reply drivers/segger/sysview.c]: done

File: drivers/segger/sysview.c (hunk 9 above)
Comment (Gary-Hobson) [reply drivers/segger/sysview.c]: done

**Comment by xiaoxiang781216** [comment]
@PetervdPerk-NXP since sched_node support the chain now, we plan to remove the hack done by https://github.com/apache/nuttx/pull/7809. Could you take a review whether the new approach satisfy your use case?

**Comment by xiaoxiang781216** [comment]
Please rebase the code

File: drivers/segger/sysview.c [hunk 10]
```diff
@@ -22,55 +22,94 @@
...
+#include <stddef.h>
+#include <syslog.h>
 #include <SEGGER_RTT.h>
 #include <SEGGER_SYSVIEW.h>
 
 #include "sched/sched.h"
 
 /****************************************************************************
- * Private Types
+ * Private Functions
```
Comment (xiaoxiang781216) [inline drivers/segger/sysview.c:39]: Private Function Prototypes

File: drivers/segger/sysview.c [hunk 11]
```diff
@@ -22,55 +22,94 @@
...
+                                  int nr, uintptr_t result);
 #endif
-};
 
 /****************************************************************************
  * Private Data
  ****************************************************************************/
 
-static struct sysview_s g_sysview =
+static struct note_driver_ops_s g_sysview_ops =
```
Comment (xiaoxiang781216) [inline drivers/segger/sysview.c:67]: add const

File: drivers/segger/sysview.c [hunk 12]
```diff
@@ -22,55 +22,94 @@
...
 #endif
 };
 
+/****************************************************************************
+ * Public Data
+ ****************************************************************************/
+
+struct note_sysview_driver_s g_sysview_driver =
+{
+    {
```
Comment (xiaoxiang781216) [inline drivers/segger/sysview.c:108]: remove two extra spaces

File: include/nuttx/segger/sysview.h [hunk 13]
```diff
@@ -25,7 +25,23 @@
...
+{
+  struct note_driver_s driver;
+  unsigned int         irq[CONFIG_SMP_NCPUS];
+};
+
+/****************************************************************************
+ * Public Data
+ ****************************************************************************/
+
+extern struct note_sysview_driver_s g_sysview_driver;
```
Comment (xiaoxiang781216) [inline include/nuttx/segger/sysview.h:44]: add ifdef/endif guard

File: drivers/segger/sysview.c [hunk 14]
```diff
@@ -365,63 +255,33 @@ void sched_note_irqhandler(int irq, FAR void *handler, bool enter)
...
-       * the active set.
-       */
-
-      if (NOTE_FILTER_SYSCALLMASK_ISSET(0, &g_sysview.syscall_marker) == 0)
-        {
-          NOTE_FILTER_SYSCALLMASK_SET(0, &g_sysview.syscall_marker);
-        }
-    }
-
+  SEGGER_SYSVIEW_NameMarker(nr, g_funcnames[nr]);
```
Comment (xiaoxiang781216) [inline drivers/segger/sysview.c:268]: it's inefficient to send the name marker every time, why remove the original logic which avoid send the name more than once.

File: include/nuttx/segger/sysview.h [hunk 15]
```diff
@@ -25,18 +25,23 @@
...
- * Pre-processor Definitions
+ * Public Types
  ****************************************************************************/
 
-#ifdef CONFIG_SEGGER_SYSVIEW_PREFIX
-#  define PREFIX(fun) sysview ## _ ## fun
-#else
-#  define PREFIX(fun) fun
-#endif
+struct note_sysview_driver_s
```
Comment (xiaoxiang781216) [inline include/nuttx/segger/sysview.h:34]: let's move to the source file and register to the global array in sysview_initialize

File: drivers/segger/sysview.c [hunk 16]
```diff
@@ -22,55 +22,94 @@
...
- * Private Types
+ * Private Functions
  ****************************************************************************/
 
-struct sysview_s
-{
-  unsigned int                 irq[CONFIG_SMP_NCPUS];
-#ifdef CONFIG_SCHED_INSTRUMENTATION_FILTER
-  struct note_filter_mode_s    mode;
+static void sysview_start(FAR struct note_driver_s *drv,
```
Comment (xiaoxiang781216) [inline drivers/segger/sysview.c:42]: let's add note_ prefix to all struct, function, variable and file name(note_sysview.c)

File: drivers/segger/sysview.c (hunk 14 above)
Comment (Gary-Hobson) [reply drivers/segger/sysview.c]: done

File: include/nuttx/segger/sysview.h (hunk 15 above)
Comment (Gary-Hobson) [reply include/nuttx/segger/sysview.h]: done

File: drivers/segger/sysview.c (hunk 16 above)
Comment (Gary-Hobson) [reply drivers/segger/sysview.c]: done

File: drivers/note/note_driver.c [hunk 17]
```diff
@@ -40,6 +40,7 @@
 #include <nuttx/note/notelog_driver.h>
 #include <nuttx/spinlock.h>
 #include <nuttx/sched_note.h>
+#include <nuttx/segger/sysview.h>
```
Comment (xiaoxiang781216) [inline drivers/note/note_driver.c:43]: remove

File: include/nuttx/segger/sysview.h [hunk 18]
```diff
@@ -25,7 +25,7 @@
  * Included Files
  ****************************************************************************/
 
-#include <nuttx/config.h>
+#include <nuttx/note/note_driver.h>
```
Comment (xiaoxiang781216) [inline include/nuttx/segger/sysview.h:28]: revert the change

File: drivers/segger/sysview.c [hunk 19]
```diff
@@ -39,47 +39,93 @@
...
+#ifdef CONFIG_SCHED_INSTRUMENTATION_SYSCALL
+  note_sysview_syscall_enter, /* syscall_enter */
+  note_sysview_syscall_leave, /* syscall_leave */
+#endif
+#ifdef CONFIG_SCHED_INSTRUMENTATION_IRQHANDLER
+  note_sysview_irqhandler,    /* irqhandler */
 #endif
 };
 
+struct note_sysview_driver_s g_note_sysview_driver =
```
Comment (xiaoxiang781216) [inline drivers/segger/sysview.c]: add static

File: drivers/segger/sysview.c [hunk 20]
```diff
@@ -333,16 +234,12 @@ void sched_note_resume(FAR struct tcb_s *tcb)
 #endif
 
 #ifdef CONFIG_SCHED_INSTRUMENTATION_IRQHANDLER
-void sched_note_irqhandler(int irq, FAR void *handler, bool enter)
+static void note_sysview_irqhandler(FAR struct note_driver_s *drv, int irq,
+                                    FAR void *handler, bool enter)
 {
-  if (!sysview_isenabled_irq(irq, enter))
```
Comment (xiaoxiang781216) [inline drivers/segger/sysview.c:338]: let's cast drv to avoid using the global variable(g_note_sysview_driver) as much as possible:
FAR struct note_sysview_driver_s *sysview = ...

File: drivers/segger/sysview.c [hunk 21]
```diff
@@ -365,63 +262,63 @@ void sched_note_irqhandler(int irq, FAR void *handler, bool enter)
...
-  if (sysview_isenabled_syscall(nr) == 0)
-    {
-      return;
-    }
-
   /* Set the name marker if the current syscall nr is not active */
 
-  if (NOTE_FILTER_SYSCALLMASK_ISSET(nr, &g_sysview.syscall_marker) == 0)
+  if (NOTE_FILTER_SYSCALLMASK_ISSET(nr,
+      &g_note_sysview_driver.syscall_marker) == 0)
```
Comment (xiaoxiang781216) [inline drivers/segger/sysview.c]: ditto

File: drivers/segger/sysview.c [hunk 22]
```diff
@@ -365,63 +262,63 @@ void sched_note_irqhandler(int irq, FAR void *handler, bool enter)
...
     {
       /* Set the name marker */
 
       SEGGER_SYSVIEW_NameMarker(nr, g_funcnames[nr]);
 
       /* Mark the syscall active */
 
-      NOTE_FILTER_SYSCALLMASK_SET(nr, &g_sysview.syscall_marker);
+      NOTE_FILTER_SYSCALLMASK_SET(nr, &g_note_sysview_driver.syscall_marker);
```
Comment (xiaoxiang781216) [inline drivers/segger/sysview.c]: ditto

File: drivers/segger/sysview.c [hunk 23]
```diff
@@ -365,63 +262,63 @@ void sched_note_irqhandler(int irq, FAR void *handler, bool enter)
...
 
       /* Use the Syscall "0" to identify whether the syscall is enabled,
        * if the host tool is closed abnormally, use this bit to clear
        * the active set.
        */
 
-      if (NOTE_FILTER_SYSCALLMASK_ISSET(0, &g_sysview.syscall_marker) == 0)
+      if (NOTE_FILTER_SYSCALLMASK_ISSET(0,
+            &g_note_sysview_driver.syscall_marker) == 0)
```
Comment (xiaoxiang781216) [inline drivers/segger/sysview.c]: ditto

File: drivers/segger/sysview.c [hunk 24]
```diff
@@ -365,63 +262,63 @@ void sched_note_irqhandler(int irq, FAR void *handler, bool enter)
...
         {
-          NOTE_FILTER_SYSCALLMASK_SET(0, &g_sysview.syscall_marker);
+          NOTE_FILTER_SYSCALLMASK_SET(0,
+            &g_note_sysview_driver.syscall_marker);
```
Comment (xiaoxiang781216) [inline drivers/segger/sysview.c]: ditto

File: drivers/segger/sysview.c [hunk 25]
```diff
@@ -365,63 +262,63 @@ void sched_note_irqhandler(int irq, FAR void *handler, bool enter)
...
 {
   nr -= CONFIG_SYS_RESERVED;
 
-  if (sysview_isenabled_syscall(nr) == 0)
-    {
-      return;
-    }
-
-  if (NOTE_FILTER_SYSCALLMASK_ISSET(nr, &g_sysview.syscall_marker) != 0)
+  if (NOTE_FILTER_SYSCALLMASK_ISSET(nr,
```
Comment (xiaoxiang781216) [inline drivers/segger/sysview.c]: ditto

Comment (Gary-Hobson) [reply drivers/segger/sysview.c]: done

File: drivers/segger/sysview.c (hunk 20 above)
Comment (Gary-Hobson) [reply drivers/segger/sysview.c:338]: done

File: drivers/segger/sysview.c (hunk 19 above)
Comment (Gary-Hobson) [reply drivers/segger/sysview.c]: done

File: include/nuttx/segger/sysview.h (hunk 18 above)
Comment (Gary-Hobson) [reply include/nuttx/segger/sysview.h]: done

File: drivers/note/note_driver.c (hunk 17 above)
Comment (Gary-Hobson) [reply drivers/note/note_driver.c]: done

File: drivers/segger/sysview.c [hunk 26]
```diff
@@ -333,16 +237,14 @@ void sched_note_resume(FAR struct tcb_s *tcb)
...
 #ifdef CONFIG_SCHED_INSTRUMENTATION_IRQHANDLER
-void sched_note_irqhandler(int irq, FAR void *handler, bool enter)
+static void note_sysview_irqhandler(FAR struct note_driver_s *drv, int irq,
+                                    FAR void *handler, bool enter)
 {
-  if (!sysview_isenabled_irq(irq, enter))
-    {
-      return;
-    }
+  FAR struct note_sysview_driver_s *driver = drv;
```
Comment (xiaoxiang781216) [inline drivers/segger/sysview.c]: need cast

File: drivers/segger/sysview.c [hunk 27]
```diff
@@ -365,63 +267,63 @@ void sched_note_irqhandler(int irq, FAR void *handler, bool enter)
...
-  if (sysview_isenabled_syscall(nr) == 0)
-    {
-      return;
-    }
-
   /* Set the name marker if the current syscall nr is not active */
 
-  if (NOTE_FILTER_SYSCALLMASK_ISSET(nr, &g_sysview.syscall_marker) == 0)
+  if (NOTE_FILTER_SYSCALLMASK_ISSET(nr,
+      &driver->syscall_marker) == 0)
```
Comment (xiaoxiang781216) [inline drivers/segger/sysview.c]: merge into one line

File: drivers/segger/sysview.c [hunk 28]
```diff
@@ -365,63 +267,63 @@ void sched_note_irqhandler(int irq, FAR void *handler, bool enter)
...
     }
 }
 #endif
 
 #ifdef CONFIG_SCHED_INSTRUMENTATION_SYSCALL
-void sched_note_syscall_enter(int nr, int argc, ...)
+static void note_sysview_syscall_enter(FAR struct note_driver_s *drv, int nr,
+                                       int argc, va_list *ap)
 {
+  FAR struct note_sysview_driver_s *driver = drv;
```
Comment (xiaoxiang781216) [inline drivers/segger/sysview.c]: add cast

File: drivers/segger/sysview.c [hunk 29]
```diff
@@ -365,63 +267,63 @@ void sched_note_irqhandler(int irq, FAR void *handler, bool enter)
...
   nr -= CONFIG_SYS_RESERVED;
 
-  if (sysview_isenabled_syscall(nr) == 0)
-    {
-      return;
-    }
-
-  if (NOTE_FILTER_SYSCALLMASK_ISSET(nr, &g_sysview.syscall_marker) != 0)
+  if (NOTE_FILTER_SYSCALLMASK_ISSET(nr,
+        &driver->syscall_marker) != 0)
```
Comment (xiaoxiang781216) [inline drivers/segger/sysview.c]: merge with previous line

File: drivers/segger/sysview.c [hunk 30]
```diff
@@ -365,63 +267,63 @@ void sched_note_irqhandler(int irq, FAR void *handler, bool enter)
...
     }
 
   SEGGER_SYSVIEW_MarkStart(nr);
 }
 
-void sched_note_syscall_leave(int nr, uintptr_t result)
+static void note_sysview_syscall_leave(FAR struct note_driver_s *drv,
+                                       int nr, uintptr_t result)
 {
+  FAR struct note_sysview_driver_s *driver = drv;
```
Comment (xiaoxiang781216) [inline drivers/segger/sysview.c]: cast

**Review by xiaoxiang781216 — APPROVED** [review]
